packages/jest/src/migrations/update-17-1-0/move-options-to-target-defaults.ts

```diff
@@ -0,0 +1,164 @@
+  if (nxJson.targetDefaults?.['@nx/jest:jest']) {
+    return;
+  }
+
+  nxJson.targetDefaults ??= {};
+
+  /**
+   * A set of targets which does not use any other executors
+   */
+  const jestTargets = new Set<string>();
+
+  const graph = await createProjectGraphAsync();
+
+  forEachExecutorOptionsInGraph(
+    graph,
+    '@nx/jest:jest',
+    (value, proj, targetName) => {
+      jestTargets.add(targetName);
+    }
+  );
+
+  // Workspace does not use jest?
+  if (jestTargets.size === 0) {
+    return;
+  }
+  // Use the project graph so targets which are inferred are considered
+  const projects = graph.nodes;
+  const projectMap = getProjects(tree);
+
+  const jestDefaults: TargetConfiguration<Partial<JestExecutorOptions>> =
+    (nxJson.targetDefaults['@nx/jest:jest'] = {});
+
+  // All jest targets have the same name
+  if (jestTargets.size === 1) {
+    const targetName = Array.from(jestTargets)[0];
+    if (nxJson.targetDefaults[targetName]) {
+      Object.assign(jestDefaults, nxJson.targetDefaults[targetName]);
+    }
+  }
+
+  jestDefaults.cache ??= true;
+
+  const inputs = ['default'];
+  inputs.push(nxJson.namedInputs?.production ? '^production' : '^default');
+  if (tree.exists('jest.preset.js')) {
+    inputs.push('{workspaceRoot}/jest.preset.js');
+  }
+  jestDefaults.inputs ??= inputs;
+
+  jestDefaults.options ??= {};
+  jestDefaults.options.passWithNoTests ??= true;
+  jestDefaults.configurations ??= {};
+  jestDefaults.configurations.ci ??= {};
+  jestDefaults.configurations.ci.ci ??= true;
+  jestDefaults.configurations.ci.codeCoverage ??= true;
+
+  // Cleanup old target defaults
+  for (const [targetDefaultKey, targetDefault] of Object.entries(
+    nxJson.targetDefaults
+  )) {
+    // If every target in every project does not use the target default anymore, it can be removed.
+    if (
+      Object.values(projects).every((p) =>
+        Object.keys(p.data?.targets ?? {}).every(
+          (targetName) =>
+            readTargetDefaultsForTarget(
+              targetName,
+              nxJson.targetDefaults,
+              // It might seem like we should use the graph here too but we don't want to pass an executor which was processed in the graph
+              projectMap.get(p.name).targets?.[targetName]?.executor
+            ) !== targetDefault
+        )
+      )
+    ) {
+      delete nxJson.targetDefaults[targetDefaultKey];
+    }
+  }
+
+  updateNxJson(tree, nxJson);
+
+  forEachExecutorOptions<JestExecutorOptions>(
+    tree,
+    '@nx/jest:jest',
+    (value, proj, targetName, configuration) => {
+      const projConfig = projectMap.get(proj);
+
+      if (!configuration) {
+        // Options
+        if (value.passWithNoTests === jestDefaults.options.passWithNoTests) {
+          delete projConfig.targets[targetName].options.passWithNoTests;
+        } else {
+          projConfig.targets[targetName].options.passWithNoTests ??= false;
```

Comment:
This could have been a different value as set in the defaults we deleted in the above loop?

```js
for (const [targetDefaultKey, targetDefault] of Object.entries(
    nxJson.targetDefaults
  )) {
  //...
}
```

Like, they could have this previously in `nx.json`:

```
{
  "targetDefaults": {
    "test": {
      "options": {
        "passWithNoTests": true
      }
  }
}
```

After this migration, it would be set to `false`, since the project did not define it previously in `project.json`, but we deleted the defaults in `nx.json`.